grails-app/controllers/org/pih/warehouse/api/InternalLocationApiController.groovy

```diff
@@ -11,30 +11,22 @@ package org.pih.warehouse.api
 
 import grails.converters.JSON
 import grails.gorm.transactions.Transactional
+
 import org.pih.warehouse.core.ActivityCode
+import org.pih.warehouse.core.InternalLocationSearchCommand
 import org.pih.warehouse.core.Location
 import org.pih.warehouse.core.LocationTypeCode
+import org.pih.warehouse.core.ReceivingBinSearchCommand
 
 @Transactional
 class InternalLocationApiController {
 
     def locationService
 
-    def list() {
-        String locationId = params?.location?.id ?: session?.warehouse?.id
-        Location parentLocation = locationId ? Location.get(locationId) : null
-        if (!parentLocation) {
-            throw new IllegalArgumentException("Must provide location.id as a request parameter")
-        }
-
-        LocationTypeCode[] locationTypeCodes = params.locationTypeCode ? params.list("locationTypeCode") : [LocationTypeCode.INTERNAL, LocationTypeCode.BIN_LOCATION]
-
-        Map activityCodeParams = [:]
-        activityCodeParams.allActivityCodes = params.allActivityCodes ? params.list("allActivityCodes") as ActivityCode[]: []
-        activityCodeParams.anyActivityCodes = params.anyActivityCodes ? params.list("anyActivityCodes") as ActivityCode[]: []
-        activityCodeParams.ignoreActivityCodes = params.ignoreActivityCodes ? params.list("ignoreActivityCodes") as ActivityCode[] : []
+    def list(InternalLocationSearchCommand command) {
```

Comment:
I didn't originally implement this with a command but `locationService.getInternalLocations` was so annoying to work with when activityCodeParams was a Map because I needed optional params and the compiler was complaining that it couldn't distinguish between methods so I got fed up enough to refactor it.